### PR TITLE
Implement dynamic schema feature for external table

### PR DESCRIPTION
**Pull request by liming01**
UPDATED:
----------
Implement dynamic schema feature for external table

Many customers complained that gptransfer will make pg_class catalog bloated. The reason
is: gptransfer will call "insert into dest_table select * from external_table", and for each dest_table, we need create one dedicated external_table because of schema difference. The dynamic schema external table can cast it's schema to any table's schema as needed.

To specify the external table be dynamic schema, add OPTIONS (dynamic_schema 'true') to 'CREATE EXTERNAL TABLE' statement. The column list definition in DDL should be empty.

A pseudo UDF GP_DYNAMIC_EXTTBL_AS_TBL('external_table', 'cast_table') is implemented to replace external table's schema by the cast_table's schema.

Signed-off-by: Tingfang Bao <bbao@pivotal.io>
---------------- Old commit message ----------------
Many customer complained that gptransfer will make pg_class catalog bloated. The reason
is: gptransfer will call "insert into dest_table select * from external_table", and for
each dest_table, we need create one dedicated external_table because of schema difference.

Here is a simple solution: each time reading the schema info of the external table, we just
read them from the dest_table in case of "insert into dest_table select * from external_table".

To specify the external table be schema-less, add OPTIONS (schemaless 'true') to 'CREATE
EXTERNAL TABLE' statement. The column list definition in DDL still kept because they will
be used in other cases.

Now only support "INSERT INTO target_table SELECT * FROM external_table", cannot support other
complex scenarios like 'INSERT INTO target_table(c1,c2) SELECT *,c1+c2 FROM external_table'.

Signed-off-by: Tingfang Bao <bbao@pivotal.io>

### CODE REVIEW

**Comment by hlinnaka** [comment]
Wow, that's a big hack. Can you come up with a more straightforward user interface for this feature? Perhaps a set-returning function, similar to dblink?

**Comment by liming01** [comment]
Hi @hlinnaka, do you mean that: we don't need the OPTIONS (schemaless 'true') for external DDL, and we only offer a set returning UDF for replacing external attribute list with specific relation's attribute list?  e.g. INSERT INTO t1 SELECT * FROM replace_attribute_list(external_table, t1);

File: src/test/regress/input/external_table.source [hunk 1]
```diff
@@ -313,8 +317,34 @@ select * from tableless_ext;
...
+format 'csv';
+create external table schemaless_ext3(a int)
+location ('file://@hostname@@abs_srcdir@/data/schemaless.csv')
+format 'csv' OPTIONS (schemaless 'true');
+
+SELECT count(*) FROM schemaless_ext1;
+SELECT count(*) FROM schemaless_ext2; -- error for the column number mismatched with file 
+SELECT count(*) FROM schemaless_ext3; -- error for the column number mismatch with file 
+
+INSERT INTO schemaless_t1 SELECT * FROM schemaless_ext1;
```
Comment (asimrp) [inline src/test/regress/input/external_table.source]: `schemaless_ext3` is the true schemaless table.  What is being tested using `schemaless_ext1` and `schemaless_ext2`?

More tests are needed here.  E.g. single transaction with multiple insert commands, each inserting into target table with distinct schema, using the same schemaless external table.
Another scenario could be:

```SQL
begin;
insert into mytable select * from schemaless;
alter table mytable <alter type/add column/drop column>;
insert into mytable select * from schemaless;
end;```

**Comment by hlinnaka** [comment]
On 08/12/17 05:51, Ming LI wrote:
> Hi @hlinnaka, do you mean that: we don't need the OPTIONS (schemaless 'true') for external DDL, and we only offer a set returning UDF for replacing external attribute list with specific relation's attribute list?  e.g. INSERT INTO t1 SELECT * FROM replace_attribute_list(external_table, t1);

Yeah, something like that.

- Heikki

File: src/test/regress/input/external_table.source (hunk 1 above)
Comment (liming01) [reply src/test/regress/input/external_table.source]: Good suggestion. I will add these scenario as your suggestion.

**Comment by liming01** [comment]
Hi @hlinnaka, 

From the user perspective, it is more friendly to use UDF, however there is a requirement I have not figure out how to implement:  external table can be specified running on master or on all segments, if we wrapper with UDF, we should also make the UDF running on master or on all segments, otherwise the performance will be greatly downgraded.

So my question is: is there any way to specify UDF to run on master or on segment?

File: src/test/regress/input/external_table.source (hunk 1 above)
Comment (adam8157) [reply src/test/regress/input/external_table.source]: It looks better to restrict schema-less external table having no schema and forbid alternating it.

**Comment by adam8157** [comment]
Besides how it looks, do we have other possible use cases of schema-less external table requiring other forms of the UI?

**Comment by hlinnaka** [comment]
On 08/12/17 11:18, Ming LI wrote:
> Hi @hlinnaka,
> 
> From the user perspective, it is more friendly to use UDF, however
> there is a requirement I have not figure out how to implement:
> external table can be specified running on master or on all segments,
> if we wrapper with UDF, we should also make the UDF running on master
> or on all segments, otherwise the performance will be greatly
> downgraded.
> 
> So my question is: is there any way to specify UDF to run on master
> or on segment?

Yes, since commit aa148d2a3c, you can specify EXECUTE ON [MASTER | ALL 
SEGMENTS] as an option to CREATE FUNCTION.


Yet another approach would be to add some special syntax for this, 
similar to the gp_dist_random pseudo-function, for example. So that you 
could do something like this:

CREATE EXTERNAL TABLE exttab () OPTIONS (dynamic_schema=true)

SELECT * FROM gp_dynamic_exttab(exttab) AS (col1 int4, col2 text);


This would be more or less the same as a UDF, perhaps exactly the same 
UI, but it might make sense to implement it as a special case in the 
parser/planner.

- Heikki

**Comment by liming01** [comment]
Thanks @hlinnaka for your helps. I will try to implement as your suggestion. ~.~

File: src/test/regress/input/external_table.source (hunk 1 above)
Comment (liming01) [reply src/test/regress/input/external_table.source]: Hi @asimrp , for your question,
> Is it possible to leverage WITH clause for schemaless option, treating it as a reloption?

I looked into the code, now in parser the statement CREATE EXTERNAL TABLE didn't include WITH clause, however it already included this ext_options_opt, so I added there for minor change. 

BTW,  from the doc perspective, the WITH clause is for storage option, while our option is only for external table, put into WITH clause may let user think this feature can works for the normal relation.

Maybe need more discussion for this issue. Thanks.

**Comment by hlinnaka** [comment]
On 15/12/17 05:24, Ming LI wrote:
>> Is it possible to leverage WITH clause for schemaless option,
>> treating it as a reloption?
> 
> I looked into the code, now in parser the statement CREATE EXTERNAL
> TABLE didn't include WITH clause, however it already included this
> ext_options_opt, so I added there for minor change.
> 
> BTW,  from the doc perspective, the WITH clause is for storage
> option, while our option is only for external table, put into WITH
> clause may let user think this feature can works for the normal
> relation.

The upcoming CREATE FOREIGN TABLE syntax will use the OPTIONS keyword:

CREATE FOREIGN TABLE [ IF NOT EXISTS ] table_name ( [
   { column_name data_type [ OPTIONS ( option 'value' [, ... ] ) ] [ 
COLLATE collation ] [ column_constraint [ ... ] ]
     | table_constraint }
     [, ... ]
] )
[ INHERITS ( parent_table [, ... ] ) ]
   SERVER server_name
[ OPTIONS ( option 'value' [, ... ] ) ]

If we're adding new syntax to CREATE EXTERNAL TABLE now, it would 
perhaps make sense to make it similar to CREATE FOREIGN TABLE.

- Heikki

**Comment by liming01** [comment]
Thanks @hlinnaka,  now I implemented as "OPTIONS (dynamic_schema 'true')", is it good enough?

**Comment by liming01** [comment]
Hi all, 

New implementation as suggestions above has push, please review it again.  Many thanks.

**Comment by adam8157** [comment]
```
FATAL:  Unexpected internal error (initsplan.c:206)
DETAIL:  FailedAssertion("!(attno >= rel->min_attr && attno <= rel->max_attr)", File: "initsplan.c", Line: 206)
HINT:  Process 20289 will wait for gp_debug_linger=120 seconds before termination.
```

Please enable assert and re-check your tests, thanks. `--enable-cassert`

**Comment by adam8157** [comment]
@hsyuan @vraghavan78 The cases fail with gporca, could you please take a look? Where is the proper place to hack? Thanks.

**Comment by liming01** [comment]
Sorry. There was some wrong, need more investigation.

**Comment by liming01** [comment]
It should run clean now. Please review it. Thanks.

**Comment by liming01** [comment]
It should run clean now. Please review it. Thanks.

**Comment by liming01** [comment]
Now icw_planner_centos6 run clean, but orca related jobs failed (icw_gporca_centos6 and icw_gporca_ubuntu16). Orca code should change correspondingly.

File: src/backend/parser/parse_relation.c [hunk 2]
```diff
@@ -899,19 +906,20 @@ parserOpenTable(ParseState *pstate, const RangeVar *relation,
...
-				   Alias *alias,
-				   bool inh,
-				   bool inFromCl)
+static RangeTblEntry *
+addRangeTableEntry_internal(ParseState *pstate,
+		RangeVar *relation,
+		Alias *alias,
+		bool inh,
+		bool inFromCl,
+		RangeVar *castRelv)
```
Comment (hsyuan) [inline src/backend/parser/parse_relation.c]: what does castRelv mean?

File: src/backend/parser/parse_relation.c [hunk 3]
```diff
@@ -938,24 +946,40 @@ addRangeTableEntry(ParseState *pstate,
...
 		inh = false;
+		if(castRelv)
+		{
+			if(!CheckDynamicOptionForExtTab(rel))
+				elog(ERROR, "Only external table with OPTIONS (%s 'true') works for %s().", DYNAMIC_EXT_TAB_OPTION, GP_DYNAMIC_EXTTAB_AS_TAB);
+
+			castRel = parserOpenTable(pstate, castRelv, NoLock, nowait, NULL);
+			rte->castRelid = castRel->rd_id;
+		}
+	}else if(castRelv)
```
Comment (hsyuan) [inline src/backend/parser/parse_relation.c]: `else if(castRelv)` in a seperate line

File: src/backend/parser/parse_relation.c (hunk 2 above)
Comment (liming01) [reply src/backend/parser/parse_relation.c]: Done. Renamed "castRelv" to "castRelrv".

File: src/backend/parser/parse_relation.c (hunk 3 above)
Comment (liming01) [reply src/backend/parser/parse_relation.c]: Done.

File: src/test/regress/input/external_table.source [hunk 4]
```diff
@@ -313,8 +322,71 @@ select * from tableless_ext;
...
+create table dyn_schema_t1 (a int, b int);
+insert into dyn_schema_t1 select a,a from generate_series(1,10) as a;
+COPY dyn_schema_t1 to '@abs_srcdir@/data/dynamic_schema.csv' CSV;
+
+-- normal external table, the same schema with dyn_schema_t1 heap table 
+create external table dyn_schema_ext1(a int, b int)
+location ('file://@hostname@@abs_srcdir@/data/dynamic_schema.csv')
+format 'csv';
+
+-- normal external table, not the same scheml with dyn_schema_t1 heap table
```
Comment (adam8157) [inline src/test/regress/input/external_table.source]: typo, scheml -> schema

Comment (liming01) [reply src/test/regress/input/external_table.source]: Fixed.

**Comment by liming01** [comment]
Hi @adam8157,  

Fixed. Also test case added. 
Thanks for your testing.

File: src/test/regress/input/external_table.source [hunk 5]
```diff
@@ -313,8 +322,72 @@ select * from tableless_ext;
...
+
+-- test: dynamic schema
+create external table dyn_schema_ext3()
+location ('file://@hostname@@abs_srcdir@/data/dynamic_schema.csv')
+format 'csv' OPTIONS (dynamic_schema 'true'); -- OK
+
+-- 2. test: select query
+SELECT count(*) FROM dyn_schema_ext1;
+SELECT count(*) FROM GP_DYNAMIC_EXTTAB_AS_TAB('dyn_schema_ext2', 'dyn_schema_t1'); -- ERROR:  Only external table with OPTIONS (dynamic_schema 'true') works for GP_DYNAMIC_EXTTAB_AS_TAB() 
+SELECT count(*) FROM GP_DYNAMIC_EXTTAB_AS_TAB('dyn_schema_ext3', 'dyn_schema_t1'); -- ok
```
Comment (yydzero) [inline src/test/regress/input/external_table.source]: Is it better to use name like 'GP_DYNAMIC_EXTTBL_AS_TBL'?  TAB is very well used for TABKey.

Comment (liming01) [reply src/test/regress/input/external_table.source]: @yydzero , good suggestion, i will correct them at the next commit. Thanks.

**Comment by hlinnaka** [comment]
Sorry for the delay.

The UI with the function is better now. Other than that, this still feels like a huge hack though. I'm not sure what exactly to suggest instead. But some thoughts:

* The use interface, with `GP_DYNAMIC_EXTTAB_AS_TAB(...)` calls, now looks like a set-returning function. Can we also implement it as a set-returning function, instead of hacking the planner?

* This basically uses the external table just as a container for connection options. The column definitions are not used. That's really hacky. We now have the concept of a FOREIGN SERVER in master. Could you use that, instead? Similar to how the dblink contrib module works (https://www.postgresql.org/docs/10/static/contrib-dblink-connect.html).

* Actually, could you just use dblink instead of all this?

* If we commit to a hack like this now, what are we going to at upgrades in the future? Are we going to add upgrade code to transform these special external tables into FOREIGN SERVERS automatically, or require that the user drop and recreate them manually? Or are we going to leave this undocumented and unsupported, except for the tool (gptransfer?) that needs this right now?

**Comment by liming01** [comment]
@hlinnaka ,  thanks for your helps.  Below is my comments for your thoughts:

>The use interface, with GP_DYNAMIC_EXTTAB_AS_TAB(...) calls, now looks like a set-returning function. Can we also implement it as a set-returning function, instead of hacking the planner?

Yes, I tried this idea at the beginning, however because we don't know the returned ROW TYPE when we define GP_DYNAMIC_EXTTAB_AS_TAB(...), we can not do it like that. Also even the func GP_DIST_RANDOM_NAME() need to change planner struct rte->forceDistRandom.

> This basically uses the external table just as a container for connection options. The column definitions are not used. That's really hacky. We now have the concept of a FOREIGN SERVER in master. Could you use that, instead? Similar to how the dblink contrib module works (https://www.postgresql.org/docs/10/static/contrib-dblink-connect.html).

FDW or dblink are a similar concept as external table, they all can access plain text as data source, however have differences:
- dblink use like UDF, and it is hard for planner to optimized the query because they are block box for query planner.
- fdw can negotiate query planning with the federation server to generate query execution plan, it has more power than others. 
- external table although access foreign data source, it's behavior is much more like an internal table (including the query planning, query executor API, statistics collection...). 

Moreover, external table only only simple text/gpfdist are supported, supporting parallel processing on segments (which has already support) is much easier.  But dblink/fdw haven't support parallel processing on segments(it means if foreign server is a MPP like gpdb, we can execute query on destination segments directly from source segment.).

In gptransfer, the reason why we use external table is to parallel processing data transfer directly from source segment (COPY ... TO... pipe) to destination segment(INSERT INTO ... SELECT * FROM external_table), meantime it can redistribute the data, which is needed if source segment number != destination segment number.

So if fdw/dblink support parallel processing, we can change gptransfer to use them. Unless gpdb will deprecate external table, add feature to external table is not conflict with fdw/dblink.

>Actually, could you just use dblink instead of all this?

Just like comments above, we need to support dblink parallel processing on segment firstly. And then change gptransfer a lot to do some preparation( e.g. create foreign server ...).

> If we commit to a hack like this now, what are we going to at upgrades in the future? Are we going to add upgrade code to transform these special external tables into FOREIGN SERVERS automatically, or require that the user drop and recreate them manually? Or are we going to leave this undocumented and unsupported, except for the tool (gptransfer?) that needs this right now?

If we want to not support external table any more, how about many application/tools developed by customer? I prefer to support them all at the same time.

**Comment by schubert** [comment]
@liming01 @yydzero @pivotal-mike @hlinnaka this discussion is outgrowing the PR - this kind of discussion needs to happen out on the dev mailing list not buried in a PR where most committers won't necessarily see it.

**Comment by yydzero** [comment]
@liming01 Would you please work with team to initiate an email to gpdb-dev alias? Thanks!

**Comment by lavjain** [comment]
+1 This looks like an excellent improvement.

**Comment by liming01** [comment]
Hi @xinzweb , 

This PR is pending. @hlinnaka think it is too hack. Do you think I need to close it without merge it?

**Comment by xinzweb** [comment]
Thanks a lot @liming01 for getting back to me. Is this one already tracked somewhere else? If that's the case, just add `ToDo` label, and this will be excluded from the PR reminder. Thanks.

**Pull request closed without merging**